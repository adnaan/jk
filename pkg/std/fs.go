package std

import (
	"os"

	"github.com/justkidding-config/jk/pkg/__std"

	flatbuffers "github.com/google/flatbuffers/go"
)

func fileInfo(path string) []byte {
	info, err := os.Stat(path)
	switch {
	case err != nil:
		return fsError(err.Error())
	case !(info.IsDir() || info.Mode().IsRegular()):
		return fsError("not a regular file")
	}
	return fileInfoResponse(path, info.IsDir())
}

func fsError(msg string) []byte {
	b := flatbuffers.NewBuilder(1024)
	off := b.CreateString(msg)
	__std.ErrorStart(b)
	__std.ErrorAddMessage(b, off)
	off = __std.ErrorEnd(b)
	__std.FileSystemResponseStart(b)
	__std.FileSystemResponseAddRetvalType(b, __std.FileSystemRetvalError)
	__std.FileSystemResponseAddRetval(b, off)
	off = __std.FileSystemResponseEnd(b)
	b.Finish(off)
	return b.FinishedBytes()
}

func fileInfoResponse(path string, isdir bool) []byte {
	b := flatbuffers.NewBuilder(1024)
	off := buildFileInfo(b, path, isdir)
	__std.FileSystemResponseStart(b)
	__std.FileSystemResponseAddRetvalType(b, __std.FileSystemRetvalFileInfo)
	__std.FileSystemResponseAddRetval(b, off)
	off = __std.FileSystemResponseEnd(b)
	b.Finish(off)
	return b.FinishedBytes()
}

func buildFileInfo(b *flatbuffers.Builder, path string, isdir bool) flatbuffers.UOffsetT {
	off := b.CreateString(path)
	__std.FileInfoStart(b)
	__std.FileInfoAddPath(b, off)
	if isdir {
		__std.FileInfoAddIsdir(b, 1)
	} else {
		__std.FileInfoAddIsdir(b, 0)
	}
	return __std.FileInfoEnd(b)
}

func directoryListing(path string) []byte {
	dir, err := os.Open(path)
	if err != nil {
		return fsError(err.Error())
	}
	infos, err := dir.Readdir(0)
	if err != nil {
		return fsError(err.Error())
	}

	b := flatbuffers.NewBuilder(1024)
	offsets := make([]flatbuffers.UOffsetT, len(infos), len(infos))
	for i := range infos {
		offsets[i] = buildFileInfo(b, infos[i].Name(), infos[i].IsDir())
	}

	__std.DirectoryStartFilesVector(b, len(offsets))
	for i := len(offsets) - 1; i >= 0; i-- {
		b.PrependUOffsetT(offsets[i])
	}
	infoVec := b.EndVector(len(offsets))

	pathOff := b.CreateString(path)
	__std.DirectoryStart(b)
	__std.DirectoryAddPath(b, pathOff)
	__std.DirectoryAddFiles(b, infoVec)
	dirOff := __std.DirectoryEnd(b)

	__std.FileSystemResponseStart(b)
	__std.FileSystemResponseAddRetvalType(b, __std.FileSystemRetvalDirectory)
	__std.FileSystemResponseAddRetval(b, dirOff)
	off := __std.FileSystemResponseEnd(b)
	b.Finish(off)
	return b.FinishedBytes()
}
