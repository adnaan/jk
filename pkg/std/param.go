package std

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/jkcfg/jk/pkg/__std"
)

func param(params Params, kind __std.ParamType, path string, defaultValue string) []byte {
	var v interface{}
	var err error

	switch kind {
	case __std.ParamTypeBoolean:
		v, err = params.GetBool(path)
	case __std.ParamTypeNumber:
		v, err = params.GetNumber(path)
	case __std.ParamTypeString:
		v, err = params.GetString(path)
	case __std.ParamTypeObject:
		// For object parameters, we merge the default value with what the user has
		// given us, which could be only a part of the default value.
		v, err = params.GetObject(path)
		if err != nil {
			break
		}
		r := strings.NewReader(defaultValue)
		// The JS side sends us JSON.
		base, _ := NewParamsFromJSON(r)
		base.Merge(v.(Params))
		v = base
	default:
		panic("param: unexpected kind")
	}

	if err != nil && strings.Contains(err.Error(), "cannot convert") {
		// TODO(dlespiau): return an error to throw a JS exception.
		fmt.Fprintf(os.Stderr, "invalid type for param '%s': %v\n", path, err)
		return []byte("null")
	} else if err != nil {
		// path not found.
		return []byte("null")
	}

	// Param returns values that can be marshalled to JSON
	data, _ := json.Marshal(v)
	return data
}
