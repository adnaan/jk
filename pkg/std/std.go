package std

import (
	"log"

	"github.com/dlespiau/jk/pkg/__std"

	flatbuffers "github.com/google/flatbuffers/go"
)

// Execute parses a message from v8 and execute the corresponding function.
func Execute(msg []byte) []byte {
	message := __std.GetRootAsMessage(msg, 0)

	union := flatbuffers.Table{}
	if !message.Args(&union) {
		log.Fatal("could not parse Message")
	}

	switch message.ArgsType() {
	case __std.ArgsWriteArgs:
		args := __std.WriteArgs{}
		args.Init(union.Bytes, union.Pos)
		write(args.Value(), string(args.Path()), args.Type())
		return nil
	default:
		log.Fatalf("unknown Message (%d)", message.ArgsType())
		return nil
	}
}
