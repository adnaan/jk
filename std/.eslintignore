flatbuffers.js
flatbuffers.d.ts
__std*.js
__std*.ts
