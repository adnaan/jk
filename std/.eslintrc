{
  "extends": "airbnb-base",
  "plugins": ["eslint-plugin-jest"],
  "globals": {
    "V8Worker2": true
  },
  "settings": {
     "import/resolver": {
        "node": {
          "paths": ["."]
        }
     }
  },
  "env": {
    "jest/globals": true
  },
  "rules": {
    "no-continue": 0,
    "prefer-const": ["error", {"destructuring": "all"}],
    "import/prefer-default-export": 0,
    "import/no-unresolved": [
      2,
      {
        "ignore": [
          "^@jkcfg/std$"
        ]
      }
    ],
    "indent": ["error", 2, {"SwitchCase": 0, "CallExpression": {"arguments": "first"}}],
    "no-restricted-syntax": ["error", "ForInStatement", "LabeledStatement", "WithStatement"],
    "no-use-before-define": ["error", { "functions": false }]
  }
}
