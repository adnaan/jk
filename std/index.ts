export { log } from './log';
export {
  stdout,
  Format,
  Overwrite,
  write,
  print,
} from './write';
export { Encoding, read, stdin } from './read';
export { parse, stringify } from './parse';
