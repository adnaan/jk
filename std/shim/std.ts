function log(value: any): void { };

enum Format {
    JSON = 1,
    YAML,
    Raw,
}

interface WriteOptions {
  format?: Format,
  indent?: number,
  override?: boolean,
}

function write(value: any, path: string, options: WriteOptions): void { };

enum Encoding {
  Bytes,
  String,
  JSON,
}

interface ReadOptions {
  format: Format,
  encoding: Encoding,
}

function read(path: string, options: ReadOptions): Promise<ArrayBuffer | string | object> { return Promise.resolve({}) };

export default {
  log, Format, write, read,
};
