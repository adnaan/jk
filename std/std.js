import { log } from 'std_log';
import { Format, write } from 'std_write';
import { Encoding, read } from 'std_read';
import { info, dir } from 'std_fs';
import { param } from 'std_param';
import { mix, patch, merge } from 'std_merge';

// The default export is deprecated and will be removed in 3.0.0.

export default {
  log,
  Format,
  write,
  Encoding,
  read,
  info,
  dir,
  param,
  mix,
  patch,
  merge,
};

export { log } from 'std_log';
export { Format, write } from 'std_write';
export { Encoding, read } from 'std_read';
export { info, dir } from 'std_fs';
export { param } from 'std_param';
export { mix, patch, merge } from 'std_merge';
