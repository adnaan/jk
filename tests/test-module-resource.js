import std from '@jkcfg/std';
import resource1 from './test-module-resource/resource';
import resource2 from './test-module-resource/submodule/resource';
import contents from './test-module-resource/fs';

Promise.all([resource1, resource2, contents]).then(resources => resources.forEach(std.log));
