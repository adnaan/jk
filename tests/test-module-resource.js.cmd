jk run -v --input-directory=/tmp %f
