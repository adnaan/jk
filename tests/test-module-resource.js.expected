{
  "module": "top"
}
{
  "module": "sub"
}
{
  "dir": {
    "files": [
      {
        "isdir": false,
        "name": "info",
        "path": "dir/info"
      }
    ],
    "name": "dir",
    "path": "dir"
  },
  "info": {
    "isdir": false,
    "name": "info",
    "path": "dir/info"
  }
}
